### ISSUE TITLE
Cannot purge; no valid URLs

### ISSUE THREAD

**Issue by ptrykulich** [workingconcept/cloudflare-craft-plugin#66]
If I change an entry through the edit page, cache will purged correctly, but if programmatically update entries and do it through a queue, I will get the error - **Cannot purge; no valid URLs**.

| | Version |
| -------- | ------- |
Cloudflare | 2.0.1
PHP version | 8.1.16
MySQL | 5.7.42
Craft Pro | 4.5.9
Yii version | 2.0.48.1

By _programmatically_, I mean that I do it using the PHP code in the module, something like this:
```
$entry = new Entry();
$entry->sectionId = 1;
$entry->typeId = 1;
$entry->authorId = 1;
$entry->title = 'Some Title';
Craft::$app->elements->saveElement($entry)
```

**Comment by bencroker** [comment]
Have you added the `Entry` element class to the `purgeElements` config setting?
https://github.com/putyourlightson/craft-cloudflare/blob/9d391f61b4b21cf354f1a3ec725be5d3bda52917/src/config.php#L38-L40

**Comment by ptrykulich** [comment]
Yes, I've added it:
<img width="278" alt="image" src="https://github.com/putyourlightson/craft-cloudflare/assets/66869015/05f62c66-80bf-4298-b77b-ef3c843ecaf6">

**Comment by bencroker** [comment]
In that case it should work. Can you verify that `Craft::$app->elements->saveElement($entry)` returns `true`?

**Comment by ptrykulich** [comment]
When I save one entry it returns true, but the thing is that I programmatically save about 4000 entries through a queue and just after that I get about 100 errors in the logs like this.

Is there some way to use something like you implemented in Blitz:

```
// Before processing the feed
Blitz::$plugin->refreshCache->batchMode = true;

// Process the feed

// After processing the feed
Blitz::$plugin->refreshCache->refresh();
```

**Comment by bencroker** [comment]
No, there is not – this plugin is nowhere near as sophisticated as Blitz.

A possible workaround would be to disable purging of entries by default, and manually call `putyourlightson\cloudflare\services\Api::purgeUrls()`, passing in the URLs of the entries saved.

**Comment by ptrykulich** [comment]
Yes, but this will disable this option for other sections. Is it possible to add specific sections instead of the element type?

**Comment by bencroker** [comment]
It’s not, but you can disable the entry element type just for the duration of your custom code.

```php
Cloudflare::$plugin->settings->purgeElements = [];

// Save elements and store URLs

Cloudflare::$plugin->api->purgeUrls($urls);
```

**Comment by ptrykulich** [comment]
Thanks! I'll try that and get back to you.

**Comment by ptrykulich** [comment]
Sorry, can you please explain? Why do we call a method to purge urls if they don't exist? I can't disable updating of entries.


```
public function purgeCachesForUrl(string $url, bool $immediately = false): void
    {
        // max limit for Cloudflare API
        $cloudflareRuleCountLimit = 30;
        $relatedRules = $this->getRulesForUrl($url);
        $urlsToPurge = [];

        foreach ($relatedRules as $rule) {
            $clearList = Json::decode($rule->urlsToClear);

            foreach ($clearList as $clearUrl) {
                $urlsToPurge[] = UrlHelper::siteUrl($clearUrl);
            }
        }

        $numRules = count($urlsToPurge);

        if ($numRules > $cloudflareRuleCountLimit) {
            throw new NotSupportedException(
                sprintf(
                    'Too many rules; API requests are limited to %d and you provided %d',
                    $cloudflareRuleCountLimit,
                    $numRules
                )
            );
        }
        // here
        if ($immediately) {
            Cloudflare::$plugin->api->purgeUrls($urlsToPurge);
        } else {
            Queue::push(new PurgeCloudflareCache(['urls' => $urlsToPurge]));
        }
    }
```

**Comment by bencroker** [comment]
What I was suggesting is that you first disable automatic purging of all element types:

```php
Cloudflare::$plugin->settings->purgeElements = [];
```

Then, after you’ve programmatically updated entries, you manually purge their URLs:

```php
Cloudflare::$plugin->api->purgeUrls($urls);
```

But now I also see what you meant about errors being logged. Are you concerned about the errors themselves or is something not working the way it should?

**Comment by ptrykulich** [comment]
I've tried your advice and it's doesn't work for me.
I am concerned about the error - _Cannot purge; no valid URLs_. It seems to me that even with the cleaning of the entry disabled, I will receive it. Maybe I'm wrong. 

I have disabled  - Automatically Purge Elements:

<img width="277" alt="image" src="https://github.com/putyourlightson/craft-cloudflare/assets/66869015/6a141e5a-2d75-4fc5-b8a0-bcb786a8d6a9">


But after resave entries, there are queue to purge urls. I see that this error occurs if the array of urls is empty.

```
public function purgeUrls(array $urls = []): mixed
    {
        if (!$this->getClient()) {
            return null;
        }

        $urls = UrlHelper::prepUrls($urls);

        // don’t do anything if URLs are missing
        if (count($urls) === 0) {
            return $this->_failureResponse(
                'Cannot purge; no valid URLs.'
            );
        }
        //...
```

In the end of  `_handleElementChange` method. There is call of method:

```
$this->rules->purgeCachesForUrl(
            $element->getUrl()
        );
```

This method calls method purgeUrls:
```
 if ($immediately) {
      Cloudflare::$plugin->api->purgeUrls($urlsToPurge);
  } else {
      Queue::push(new PurgeCloudflareCache(['urls' => $urlsToPurge]));
  }
```       
End if I have empty rules, or url which is not related to rule, this method will call with empty urls, and it will give the error.

**Comment by bencroker** [comment]
I believe that error message is just being incorrectly logged, there’s nothing to worry about and I’ll add a fix for this.

But I’m confused as to whether otherwise everything is working for you or not? This issue is getting very long and I’m not sure where we are with it.

**Comment by ptrykulich** [comment]
Yes, everything else works, thanks. I was just interested in this particular error.

**Comment by bencroker** [comment]
Great! Fixed in https://github.com/putyourlightson/craft-cloudflare/commit/763e66f366cdbd89aa92130af78276c39b835b83 and released in [2.0.2](https://github.com/putyourlightson/craft-cloudflare/releases/tag/2.0.2).

**Comment by ptrykulich** [comment]
Thank you so much!
